### ISSUE TITLE
Intermittent 502 Bad Gateway error for SQS

### ISSUE THREAD

**Issue by davidgoate** [localstack/localstack#966]
# Environment:

* Docker image: `localstack/localstack:latest (c6f4c8405080)`
* OSX 10.14

# Issue

Intermittently I get this error when running my Java based tests against localstack.

>Factory method 'amazonSQS' threw exception; nested exception is com.amazonaws.services.sqs.model.AmazonSQSException: null (Service: AmazonSQS; Status Code: 502; Error Code: 502 Bad Gateway; Request ID: null)

I use a junit docker rule like this:

```xml
<dependency>
            <groupId>com.arakelian</groupId>
            <artifactId>docker-junit-rule</artifactId>
            <version>2.2.2</version>
            <scope>test</scope>
</dependency>
```

```java
public class LocalStackDockerRule extends DockerRule {

    private static final String DOCKER_IMAGE = "localstack/localstack:latest";
    private static final String CONTAINER_NAME_PREFIX = "localstack-";

    public LocalStackDockerRule(final PortBinding binding, final String services, final String portToTest) {
        super(ImmutableDockerConfig.builder()
            .name(CONTAINER_NAME_PREFIX + binding.hostPort())
            .image(DOCKER_IMAGE)
            .ports(portToTest)
            .alwaysRemoveContainer(true)
            .addContainerConfigurer(c -> c.env("SERVICES=" + services, "HOSTNAME_EXTERNAL=" + binding.hostIp()))
            .addHostConfigurer(bindPorts(getPortBindings(binding, portToTest)))
            .addStartedListener(waitForContainerToBeReady(portToTest))
            .addStartedListener(c -> c.waitForLog("Ready."))
            .build());
    }

    private static Map<String, List<PortBinding>> getPortBindings(final PortBinding binding, final String portToTest) {
        final Map<String, List<PortBinding>> bindings = new HashMap<>(1, 1);
        bindings.put(portToTest, singletonList(binding));
        return bindings;
    }

}
```

The main point being that the code is (at least in theory) supposed to wait for the SQS port (`4576`) to respond and for the message `Ready.` to appear in the container logs. 

Is it somehow possible that even after these two events localstack isn't quite ready? Under what conditions would localstack return a HTTP 502?

**Comment by Pedrock** [comment]
I'm having the same issue.

**Comment by astronouth7303** [comment]
I'm having a similar problem using the `localstack/localstack` docker hub image as a service on GitLab CI, testing with python and `localstack-client`.

**Comment by davidgoate** [comment]
@Pedrock and @astronouth7303 do you only get this shortly after the image has started or also if the image has been running a while? I only (so far) see this when the image has not been running for very long.

**Comment by astronouth7303** [comment]
Yeah, I figured out it's only right after the image has started.

I added a similar python script to block until SQS is available.

**Comment by johnmphillips** [comment]
I'm also having the same issue, I implemented a similar workaround, except it never seems to become ready, even after 5 minutes of trying

**Comment by Pedrock** [comment]
@johnmphillips It should only take a second or so. You probably have a different problem.

**Comment by Musikolo** [comment]
Yes, the workaround for me is adding a delay of 1 second once `LocalStack` is supposed to be ready. A very annoying bug that forces us to do nasty hacks like this. I hope it gets fixed shortly...

**Comment by monkut** [comment]
seeing something similar with only the s3 service....

**Comment by fsw0422** [comment]
This has to be fixed... This causes our integration test to be super flaky. We create the queue with the init script as the document says by mounting the shell script but this sometimes fails...

**Comment by Pedrock** [comment]
Let's not forget this is an open source project. Nobody is obligated to fix this and anybody can fix it by submitting a PR.

**Comment by fsw0422** [comment]
@Pedrock had no intention of forcing anyone. Maybe I can take a look for this

**Comment by PyvesB** [comment]
@CandleCandle and I were also facing similar 502 Bad Gateway issues when using the SQS service.

Underneath the hood, Localstack's SQS service fires up a Java process, using _elasticmq-server.jar_. Some of the environment variables which were being picked up by the Java process were causing it to fail at startup. In our case, launching Localstack with `DEBUG=1` helped us track down the problem, as this printed the stacktraces generated by _elasticmq-server_.

Hope this helps others get to the bottom of these error responses!

**Comment by whummer** [comment]
Thanks for reporting @davidgoate @Pedrock @fsw0422 @astronouth7303 @johnmphillips @Musikolo @monkut @PyvesB . 

This should (hopefully) be fixed in #1807 - this PR adds a health check that polls the status of SQS and waits until it becomes available at startup.

Closing this issue - please re-open this issue if the problem persists. Thanks!
